### ISSUE TITLE
NotifiableError typed to any

### ISSUE THREAD

**Issue by toxaq** [bugsnag/bugsnag-js#641]
Hi, I was just wondering why NotifableError is typed to any when it doesn't accept any?

I naively passed in an unknown response object to `client.notify()` and get no errors in typescript yet in production I get the not so helpful `Bugsnag usage error. notify() expected error/opts parameters, got unsupported object`. This is exactly the sort of error typescript is great for avoiding. If the any type was removed from the below I would've got this right first time:

     export type NotifiableError = Error
      | { errorClass: string; errorMessage: string; }
      | { name: string; message: string; }
      | any;

Am I missing something? It seems like dropping any would resolve the runtime issue by catching it at compile time.

**Comment by mattdyoung** [comment]
Hi @toxaq 

We do still need to allow `any` for reasons discussed here:
https://github.com/bugsnag/bugsnag-js/issues/278#issuecomment-349570427

But as you can see we add more specific definitions before `any` for type-hinting.

Thanks!

**Comment by toxaq** [comment]
Hi @mattdyoung, appreciate the reply. I don't know if there's something wrong with my tsconfig then because `any` immediately invalidates the prior "types". Even using the `name` `message` fields gives no assistance as the moment you type anything else it assumes it's `any`. There's no warning, no type assistance and errors in production. Seems a strange way of handling things. 

As the library I'm using is a closed 3rd party product, there's little way of finding out what it's giving as an error response until it's happened. Of course then finding out bugsnagjs can't handle it and won't give me any information at all doesn't seem ideal. Is there a reason that an unknown object couldn't be JSON stringified and dumped into an unknown error code (and this with further info) rather than the less helpful `got unsupported object`?

**Comment by bengourley** [comment]
Hi @toxaq, `@bugsnag/js` maintainer here 👋 

When we added the union type for `NotifiableError`, the expected UX in VSCode was that when you start typing `bugsnagClient.notify(…` you'd see the types suggested to you in the order they are defined.

I just checked this out and it seems like `any` takes precedence over all the other values and it sounds like this matches up with what you're seeing:

![image](https://user-images.githubusercontent.com/609579/68758286-6c4fe580-0605-11ea-9d74-006b5297fbdc.png)

> As the library I'm using is a closed 3rd party product, there's little way of finding out what it's giving as an error response until it's happened.

Is the library you're using typed? I think you would still need to know the structure of its errors to know what to pass on to Bugsnag.

Ultimately we do want to allow¹ anything here in JS and we'll try to do the best thing, but maybe it's reasonable to force a TypeScript user to cast an unknown thing to a list of known types that `notify()` can do something with.

¹allow as in tolerate – the best thing to do is always pass an `Error` (or subclass of).

I think the resulting `NotifiableError` type would look something like this:

```ts
type NotifiableError = Error
  | { errorClass: string; errorMessage: string }
  | { name: string; message: string }
  | string
  | number
  | boolean
  | function
```

As to your suggestion:

> Is there a reason that an unknown object couldn't be JSON stringified and dumped 

We [already do this for [automatically detected unhandled promise rejections](https://github.com/bugsnag/bugsnag-js/blob/cd0b919793dfe07402e40216b935725ac68faa87/packages/plugin-window-unhandled-rejection/unhandled-rejection.js#L37-L47), and so we could apply this to more kinds of error.

I'll reopen this issue, because the type of `NotifiableError` is not doing what we expected.

**Comment by toxaq** [comment]
Hi @bengourley ! Appreciate the response.

Old school Visual Studio C# dev here but since moved on to OSX now with typescript. So I appreciate the benefits of intellisense autofill but the real benefit to me is the typescript compiler, so dropping `any` would've flagged my runtime issue well before finding the `unsupported object` logs in bugsnag.

The 3rd party library is precompiled minified javascript, so whilst I could trawl through it, all I know is it returns an error object, just not an actual `Error` object. So for me the ideal solution would be that bugsnag would fall back to stringifying the unknown object and would be happy if the error name was `received unknown object` and the message was the object values. That way with the information of the object I could better handle it and even write some tests based on an exact object.

Thanks for taking the time to look at this.

**Comment by xander-jones** [comment]
Hi @toxaq, a fix for this was released in [v7.0.0](https://github.com/bugsnag/bugsnag-js/releases/tag/v7.0.0)
